src/Pos/Block/Logic/Internal.hs

```diff
@@ -79,17 +82,29 @@ type MonadBlockBase ssc ctx m
        -- Needed by some components.
        , MonadGState m
        -- This constraints define block components' global logic.
-       , HasLens TxpGlobalSettings ctx TxpGlobalSettings
        , HasLens LrcContext ctx LrcContext
+       , HasLens TxpGlobalSettings ctx TxpGlobalSettings
        , SscGStateClass ssc
+       , HasLens GenesisStakeholders ctx GenesisStakeholders
+       , MonadDelegation ctx m -- TODO: split dlgApplyBlocks blocks
        , MonadReader ctx m
        )
 
 -- | Set of constraints necessary for high-level block verification.
 type MonadBlockVerify ssc ctx m = MonadBlockBase ssc ctx m
 
 -- | Set of constraints necessary to apply or rollback blocks at high-level.
+-- Also normalize mempool.
 type MonadBlockApply ssc ctx m
+     = ( MonadBlockApplyDB ssc ctx m
+       -- Needed for normalization.
+       , MonadTxpMem TxpExtra_TMP ctx m
+       , SscLocalDataClass ssc
+       , HasLens UpdateContext ctx UpdateContext
+       )
+
+-- | Set of constraints necessary to apply or rollback blocks at high-level (without normalization).
+type MonadBlockApplyDB ssc ctx m
```

Comment:
Reverse definition order is posos. I'd swap `MonadBlockApplyDB` and `MonadBlockApply`. And rename, yes.